lib/charms/prometheus_k8s/v0/prometheus.py

```diff
@@ -684,6 +794,77 @@ def _set_unit_ip(self, event):
                 self._charm.model.get_binding(relation).network.bind_address
             )
 
+    def _label_alert_topology(self, rule):
+        """Insert juju topology labels into an alert rule.
+
+        Args:
+            rule: a dictionary representing a Prometheus alert rule.
+
+        Returns:
+            a dictionary representing Prometheus alert rule with Juju
+            topology labels.
+        """
+        metadata = self._scrape_metadata
+        labels = rule.get("labels", {})
+        labels["juju_model"] = metadata["model"]
+        labels["juju_model_uuid"] = metadata["model_uuid"]
+        labels["juju_application"] = metadata["application"]
+        rule["labels"] = labels
+        return rule
+
+    def _label_alert_expression(self, rule):
+        """Insert juju topology filters into a Prometheus alert rule.
+
+        Args:
+            rule: a dictionary representing a Prometheus alert rule.
+
+        Returns:
+            a dictionary representing a Prometheus alert rule that filters based
+            on juju topology.
+        """
+        metadata = self._scrape_metadata
+        topology = (
+            'juju_model="{}", juju_model_uuid="{}", juju_application="{}"'.format(
+                metadata["model"], metadata["model_uuid"], metadata["application"]
+            )
+        )
+        expr = rule["expr"]  # a rule has to have an "expr"
+        expr = expr.replace("%%juju_topology%%", topology)
```

Comment:
Is there a generic way to add a topology filter to a rule, to obviate "%%juju_topology%%"?

Reply:
Not sure what you mean by "generic way". A PromQL parser would be pretty generic.